miniapp/miniapp_communication.cpp

```diff
@@ -0,0 +1,610 @@
+}
+
+std::string string_full(std::string name, Device D, Device D_comm,
+                        RequireContiguous require_contiguous) {
+  std::stringstream s;
+  s << name << " " << D << ":" << D_comm;
+  if (require_contiguous == RequireContiguous::Yes)
+    s << " contiguous    ";
+  else
+    s << " non-contiguous";
+  return s.str();
+}
+
+std::string string_full(std::string name, Device D, Device D_comm,
+                        RequireContiguous require_contiguous_send,
+                        RequireContiguous require_contiguous_recv) {
+  std::stringstream s;
+  s << name << " " << D << ":" << D_comm;
+  if (require_contiguous_send == RequireContiguous::Yes)
+    s << " contiguous-send    ";
+  else
+    s << " non-contiguous-send";
+  if (require_contiguous_recv == RequireContiguous::Yes)
+    s << " contiguous-recv    ";
+  else
+    s << " non-contiguous-recv";
+  return s.str();
+}
+
+void output(int64_t run_index, std::string name, double elapsed_time, int rank, const Options& opts,
+            Size2D grid_size) {
+  if (rank == 0 && run_index >= 0) {
+    std::cout << "[" << run_index << "]"
+              << " " << name << " " << elapsed_time << "s"
+              << " " << dlaf::internal::FormatShort{opts.type} << " " << opts.m << " " << opts.mb << " "
+              << grid_size << " " << pika::get_os_thread_count() << " " << opts.backend << std::endl;
+  }
+}
+
+template <class CommPipeline, class T, Device D, class F>
+void benchmark_rw(CommPipeline& pcomm, Matrix<T, D>& matrix, F&& schedule_comm_function) {
+  for (auto& index : iterate_range2d(matrix.distribution().local_nr_tiles())) {
+    start_detached(schedule_comm_function(pcomm.exclusive(), matrix.readwrite(index)));
+  }
+}
+template <class CommPipeline, class T, Device D, class F>
+void benchmark_ro(CommPipeline& pcomm, Matrix<T, D>& matrix, F&& schedule_comm_function) {
+  for (auto& index : iterate_range2d(matrix.distribution().local_nr_tiles())) {
+    start_detached(schedule_comm_function(pcomm.exclusive(), matrix.read(index)));
+  }
+}
+template <class CommPipeline, class T, Device D_in, Device D_out, class F>
+void benchmark_ro_rw(CommPipeline& pcomm, Matrix<T, D_in>& matrix_in, Matrix<T, D_out>& matrix_out,
+                     F&& schedule_comm_function) {
+  for (auto& index : iterate_range2d(matrix_out.distribution().local_nr_tiles())) {
+    start_detached(schedule_comm_function(pcomm.exclusive(), matrix_in.read(index),
+                                          matrix_out.readwrite(index)));
+  }
+}
+
+template <Backend B, class CommPipeline, class T>
+void benchmark_all_reduce(int64_t run_index, const Options& opts, Communicator& world,
+                          CommPipeline&& pcomm, Matrix<const T, Device::CPU>& matrix_ref) {
+  using dlaf::comm::scheduleAllReduce;
+
+  constexpr Device D = DefaultDevice_v<B>;
+  Matrix<T, D> matrix = get_matrix_on_device_sync<D>(matrix_ref);
+  Matrix<T, D> matrix_out = get_matrix_on_device_sync<D>(matrix_ref);
```

Comment:
Is the reason for preferring `get_matrix_on_device_sync` over `MatrixMirror` that you actually want to force `matrix` and `matrix_out` to be new matrices, independent of the device?

Reply:
Exactly, I want to avoid an extra CPU copy if the resulting matrix should be on GPU.

Reply:
I clarified this directly with @rasolca and this is a combination of:
- Always wanting to work with a copy of the reference, and
- Wanting to avoid an unnecessary copy in the GPU case.

The reference is always on CPU. With `MatrixMirror` one would have first one copy of the reference to CPU, and then another copy from CPU to GPU. The CPU-side copy created for `MatrixMirror` is not used. With `get_matrix_on_device_sync` one can copy directly from the reference to the GPU.

On CPU there's no difference.

Thanks for the explanation @rasolca!